Various small cleanups in auxpow code

domob1812 opened the pull request:
This is a set of smaller but IMHO straight-forward clean-ups and improvements to test coverage of the auxpow code.

domob1812 commented:
@JeremyRand FYI:  I've added two more commits that further clean up `test_auxpow.py`.  I guess that's ok, since you haven't yet looked at the code.  (And if you have already, you can just look at the two new commits.  The previous ones are not changed at all.)

domob1812 commented:
@JeremyRand, please take a quick look at this and my other PRs.  All of them should be straight-forward to review.  Thanks!

JeremyRand commented:
Sigh, it appears that Travis CI has flagged all of your PR's (and quite possibly this entire repo) as abusive.  I assume this is a false positive and not any intentional malice on their end (it seems to have happened to 2 of Namecoin's other repos before).  I've just contacted Travis CI support to see if they can fix the issue, so that Travis can actually run the necessary checks on your PR's.  The last time this happened (February 2018) they resolved it within around 3.5 days.

Anyway, I'll try to do some manual review of your PR's before Travis fixes their issue, but a merge probably won't occur until Travis is fixed.

domob1812 commented:
Ok, thanks.  Let's hope they get their stuff fixed soon-ish.  Can you not just run `pytest` manually in case they take a long time?  What other tests are there?

JeremyRand commented:
ACK 0b7dba07a768e3bb5d69f7ce8dba89bbf051f6d0.  I have not tested the code, but I have reviewed it and it looks OK, I agree it can be merged, subject to a Travis check.  I haven't yet reviewed the other commits.

JeremyRand commented:
@domob1812 Speaking as someone who's reading these tests for the first time in several months, I found the original comments ("We can't pass the real height because it's below a checkpoint, and the deserializer expects ElectrumX to strip checkpointed AuxPoW.") to be much less confusing than the comments in 826a1bc7f46b3f39376be6a11d56355fe37f7e17 ("This takes care of faking the block height so that we always deserialize it with auxpow (bypassing the checkpoint test).").  In particular, it's not clear in the new comments what the "checkpoint test" does or why we're faking the height to bypass it.  Also I'd prefer that "auxpow" be capitalized as "AuxPoW" in the comments since that seems to be more consistent with the rest of the comments.

ACK 826a1bc7f46b3f39376be6a11d56355fe37f7e17.  I have not tested the code, but I have reviewed it and it looks OK, I agree it can be merged, subject to a Travis check and the above requested changes. I haven't yet reviewed the other commits.

JeremyRand commented:
@domob1812 The name of the exception in 90c89dcc4d62878746072a7ceee4602cd204349e is inconsistent with the other exceptions.  Specifically, the existing convention is to say `Root` rather than `Commitment`, and have `Root` directly following `Coinbase`.  So I'd suggest naming it `AuxPoWCoinbaseRootMissingError`.

Similarly, I'd suggest renaming `test_rejects_invalid_auxpow_commitment` to `test_should_reject_coinbase_root_missing`.

Also it's not immediately obvious what the purpose is of setting the timestamp to 42.

domob1812 commented:
Thanks for the review, I've addressed all your comments.  To me, "coinbase root" still does not mean anything, while it is clear what the "commitment" is (if one knows how auxpow works).  So while my name was not in line with the existing ones, I think it was better (i.e. the existing convention was not ideal), but I'm fine going with the existing code on this.

JeremyRand commented:
Triggering a Travis rebuild....

JeremyRand commented:
ACK 41d4d7ff929e13c3ce0768349091eb1465162478.  I have not tested the code, but I have reviewed it and it looks OK, I agree it can be merged.  I haven't yet reviewed the later commits.

JeremyRand commented:
ACK 60e5b6dd2141b1d55b5f02fada04c44c4fb1e314.  I have not tested the code, but I have reviewed it and it looks OK, I agree it can be merged. I haven't yet reviewed the later commits.

JeremyRand commented on electrum/tests/test_auxpow.py:
```diff
@@ -37,61 +69,52 @@ def test_deserialize_auxpow_header_explicit_coinbase(self):
         self.assertEqual(expected_coinbase_txid, observed_coinbase_txid)
 
         coinbase_merkle_branch = header_auxpow['coinbase_merkle_branch']
```
It looks like this PR removes all code that uses the `coinbase_merkle_branch` variable.  If that's accurate, I think this variable should be removed as well since it's unused.

domob1812 replied:
Indeed, good catch!  I've removed the unused variable.

JeremyRand commented on electrum/tests/test_auxpow.py:
```diff
@@ -37,61 +69,52 @@ def test_deserialize_auxpow_header_explicit_coinbase(self):
         coinbase_merkle_index = header_auxpow['coinbase_merkle_index']
         self.assertEqual(0, coinbase_merkle_index)
 
         chain_merkle_branch = header_auxpow['chain_merkle_branch']
```
It looks like this PR removes all code that uses the `chain_merkle_branch` variable. If that's accurate, I think this variable should be removed as well since it's unused.

domob1812 replied:
Also removed this one.

JeremyRand commented:
ACK bebaeb1efbd8131324c524069316a499db59c343 subject to the above requested changes; I have not tested the code, but I have reviewed it and it looks OK (subject to the above requested changes), I agree it can be merged. I haven't yet reviewed the later commits.

JeremyRand commented:
ACK f66527623c64a22ef3955478fbb950a6d91bcc85 subject to the above requested changes; I have not tested the code, but I have reviewed it and it looks OK (subject to the above requested changes), I agree it can be merged.

domob1812 commented:
Thanks for the reviews!  I've addressed your comments, so all should be fine.

In case Travis is fine and because you already reviewed the change, do you want me to merge this (and future similar PRs) by myself or do you want to keep control over the merges?  I'm fine with whatever is preferrable for you.

JeremyRand commented:
Sigh, the fine people (/sarcasm) at Travis have flagged us for abuse yet again, just a few days after they told me they had cleared the abuse flag.  I've just sent them a complaint, hopefully they'll get it resolved quickly.

In any event, in cases where both of the following are true:

1. Travis succeeds.
2. I've ACKed the PR.

Then you're welcome to merge.  Note however that merging is a bit tricky: we need to use the Bitcoin Core GitHub merge script, but a few lines of that script need to be commented out (IIRC it's the lines that deal with symlinks; this repo has symlinks but Bitcoin Core's script errors when it sees symlinks).  I can post my edited version of the script somewhere if you want it.

domob1812 commented:
Ok, then let's hope they fix it again.  I guess it is fine for me to wait until you merge, before I mess things up. :)

(On our Xaya projects (except Core), what I do for merges is simply use Github, then check those commits out in my local repository, compare to my local branch, and amend the merge commit with GPG signature if there are no diffs.  But that's a bit hacky.)

JeremyRand commented:
Triggering a rebuild...

JeremyRand commented:
According to Travis CI support:

> Hey Jeremy,
> 
> Thank you for your reply and I'm sorry for the continued troubles.
> 
> I believe this time it was specific to the `domob1812` user who made the commits.
> 
> I've proceeded with unflagging their account as well. Can you ask them to try again?
> 
> Thank you for your understanding and have a great day!

JeremyRand commented:
ACK 47af10e6c4853cadbf751006426f36a31e4178cb; I have not tested the code, but I have reviewed it and it looks OK, I agree it can be merged.

domob1812 commented:
Thanks for the review, merge and follow-up with Travis!  Then let's hope that the issues are fixed in the future for me as well. :)
